### PR TITLE
syncthing: Avoid syncing directories masked by native volumes

### PR DESCRIPTION
**Pull request by cg505**
Now that I've rewritten this three times I really hope it works.

This allows you to specify your working directory as a bind volume, but specify
./node_modules as a native volume, to avoid having to sync all the changes to
that directory back and forth from your laptop to your sandbox.

The logic is currently best-effort. It notices common cases that allow
subdirectories to be ignored during syncing, but it does not guarantee that it
will notice if a directory can be ignored. This is to avoid ridiculous code
complexity and/or a total rewrite of how we collapse syncthing mounts.

In an ideal world, we would rewrite that code. Since I thought about it a fair amount, I will give what I think is a better solution, for future reference. This PR does NOT implement this approach.
The `Mount` struct becomes a tree, defined as such:
```go
type Mount struct {
	// Path is a full path for the root of the tree, and a relative path for children.
	Path string
	// SyncAll specifies that all files should be synced, but this can be overridden by children.
	SyncAll bool
	// Ignore specifies that this directory or file should not be synced. Ignore should not be true at the root of the tree.
	Ignore bool
	// Children allow for Ignore or SyncAll to be overridden for specific subpaths. Children should be directly beneath the Path in the filesystem (not multiple subdirectories below) to make it easier to merge trees.
	Children []Mount
}
```
It should be fairly straightforward to see how to convert `BindVolume`s into these new `Mount`s. Then, collapsing mounts becomes an issue of merging trees, which can follow relatively simple rules (e.g. `SyncAll` overrides `Ignore` at the same level, `Ignore` children are only kept if they are present in both trees, etc). If you want to do this, good luck!

### CODE REVIEW

File: pkg/syncthing/client.go [hunk 1]
```diff
@@ -326,6 +378,10 @@ func (c Client) WriteConfig(idPathMap map[string]string) error {
 		}
 
 		stignore, ok := m.GetStignore()
+		log.WithField("stignore", stignore).WithField("ok", ok).Info("GetStignore")
```
Comment (kklin) [inline pkg/syncthing/client.go]: Leftover from debugging?

File: cli/up/up.go [hunk 2]
```diff
@@ -400,17 +400,49 @@ func startTunnel(ncc node.ControllerClient, token, name, hostIP string,
...
+			if v.Type == composeTypes.VolumeTypeBind {
+				continue
+			}
+
+			// Search through existing mounts for this service to see which need
+			// Masks.
+			for bindTarget, volume := range bindVolumes {
+				relPath, err := filepath.Rel(bindTarget, v.Target)
+				if err != nil || strings.HasPrefix(relPath, "..") {
+					// This native volumes does not fall under the bind volume.
```
Comment (kklin) [inline cli/up/up.go]: s/volumes/volume

File: pkg/syncthing/client.go [hunk 3]
```diff
@@ -107,25 +117,26 @@ func (m Mount) GetStignore() (stignore string, needed bool) {
...
 	return hash.DNSCompliant(m.Path)
 }
 
-// Syncs returns whether the mount already syncs the given child.
-func (m Mount) Syncs(child string) bool {
-	if m.SyncAll {
-		return true
-	}
-
+// Includes returns whether the mount already include the given child.
```
Comment (kklin) [inline pkg/syncthing/client.go]: s/include/includes

File: pkg/syncthing/client.go [hunk 4]
```diff
@@ -188,13 +209,45 @@ func NewClient(volumes []string) Client {
...
+					} else if mount.SyncAll {
+						// Look for a matching ignore in the child.
+						for _, mountIgnore := range mount.Ignore {
+							if mountIgnore == relIgnore {
+								// We found a match!
+								newIgnore = append(newIgnore, parentIgnore)
+								break
+							}
+						}
+					}
```
Comment (kklin) [inline pkg/syncthing/client.go:242]: Logging any ignores that we're dropping would be helpful I think. Could just be at debug.

File: pkg/syncthing/client.go [hunk 5]
```diff
@@ -188,13 +209,45 @@ func NewClient(volumes []string) Client {
...
+				}
+				parent.Ignore = newIgnore
+
+			case parent.Includes(relPath):
+				// This mount will already be included. Ignores may be
+				// discarded. Too bad.
```
Comment (kklin) [inline pkg/syncthing/client.go:248]: Same here re logging that we're dropping the ignores

**Review by kklin — APPROVED** [review]
Just left some small comments

File: pkg/syncthing/client.go (hunk 4 above)
Comment (cg505) [reply pkg/syncthing/client.go:242]: There isn't an easy way for us to tell whether the drops were necessary or not here. We could just log each one we drop, but we would end up logging the drops even when it was the correct thing to do. I think we should just leave this.

File: pkg/syncthing/client.go (hunk 5 above)
Comment (cg505) [reply pkg/syncthing/client.go:248]: Will do here.

File: pkg/syncthing/client.go [hunk 6]
```diff
@@ -206,6 +266,11 @@ func NewClient(volumes []string) Client {
 					// include list.
 					parent.Include = append(parent.Include, relPath)
 				}
+				if len(mount.Ignore) > 0 {
+					// It is too complicated to calculate which ignores can be
+					// kept.
+					log.WithField("Ignore", mount.Ignore).Debug("Dropping child ignores from updated parent mount.")
```
Comment (cg505) [inline pkg/syncthing/client.go:272]: added this one too

**Pull request merged**